source/common/http/http1/codec_impl.cc

```diff
@@ -372,8 +372,9 @@ void RequestEncoderImpl::encodeHeaders(const RequestHeaderMap& headers, bool end
   const HeaderEntry* host = headers.Host();
   bool is_connect = HeaderUtility::isConnect(headers);
 
-  // Headers must be present. This would only happen from downstream traffic (in which case
-  // downstream codecs would reject) or a faulty filter (in which case the router would reject).
+  // Headers must be present. If downstream traffic is missing headers, downstream codecs will
+  // reject the request. If a faulty filter removes these headers, the router will reject before
+  // forwarding the request headers here.
   ASSERT(method, ":method must be specified.");
   ASSERT(path || is_connect, "path must be specified");
   ASSERT(host || !is_connect, "host must be specified for a CONNECT request");
```

Comment:
Should we just assert checkHeaderMap?
Also maybe update encodeHeaders include comments about this requirement?

Reply:
Also are we comfortable with the router being the terminal filter?  I know it is today but I thought once we add upstream filters that won't be the case any more.  I wonder if we should just handle errors here and return a status or something?   cc @snowp

Reply:
I think it would still be the terminal filter as far as the HCM is concerned, but it definitely wouldn't be the last extension point that could be modifying the headers before we hit the upstream codec. I haven't looked at this PR in detail, but conceptually it seems like doing this validation makes more sense in the codec?

Reply:
Oh! Yes. I could change `encodeHeaders` to return a status that could be handled in the caller's code. 

Doing it in the codec would be a lot cleaner in terms of extensibility and not worrying about the problem happening later. What worried me was how large-scale the change would be if `encodeHeaders` returned a status. I'll give it a try now, and see if I hit bad trouble.

Reply:
I wonder if we may run into the problem where a buggy filter removes a required header and then subsequent filter in the chain crashes because it expects the header to be in the map. Would it make sense to check the integrity of the header map after calling `decodeHeaders()` for each filter and abort with 500 if a filter messes up the header map?

Reply:
That's also true. I'm not sure what the further extension points would be but at this point it seems like there are many places that can modify headers and break later processing.

At this point it seems like after codecs decode, we know
(1) For response we have a Status
(2) For request we have method. path and scheme for non-connect, and host/authority for connect or h/1. (https://tools.ietf.org/html/rfc7540#section-8.1.2.3, https://tools.ietf.org/html/rfc7540#section-8.3, https://tools.ietf.org/html/rfc7230#section-5.4)

Should HeaderMap treat these headers specially and skip their removal (as well as skip removal in HeaderMap::remove/removeIf/removePrefix)?

Reply:
I think filters are the only place at this point that can modify header map, such that it is missing (or have invalid values) for critical headers. You've listed out the headers without which codecs will reject requests or responses without before filter processing. So i think we just need graceful handling of misbehaving filters that mess with these headers.
From earlier Slack discussion we decided that preventing removal (or invalid modification) of the critical headers will add too much complexity and thought that either adding defaults or aborting request with 500 were good options.
My choice would be to abort with 500, mostly because it is not clear to me what the good default should be.
The reason I think we should check consistency of the header map after invoking decode/encodeHeaders() of filters are WASM filters, as I think we these might be of lower quality than current C++ filters.

Reply:
Resolution: Fixed with per-iteration check in the filter manager. Avoid the issue of relying on router filter being terminal. FM sends back direct response and debug logs point to which filter made the error.

@snowp: I don't think there's a way to get the codec to do these checks. The codec has no way of handling the error at that point. I think future extensions that may end up manipulating need to harden against causing errors as well.

re original comment: encodeHeaders doc string updated, ASSERT on required header check.

Reply:
Hm, I mildly prefer having this function do the checks, and return a failure status, rather than having the caller have to know to do the check beforehand.  What do folks think about that?

Reply:
I think the concern about doing the check here is it will make the change much larger. In principle I agree it would be better, but this seems fine to me also. I'm fine either way.

Reply:
I could build it off as a starting point, my concern is that every caller of `encodeHeaders` would need to handle the status (if absl::Status is used, whose result can't be ignored), and I'm not sure if it's worth implementing in every place. I could start with `IgnoreError()` calls for now though. I do think the change is better done inside, especially if there are other places that will modify headers before encoding.

Reply:
Tracing through the change now. I'll push the change as soon as I can get tests working, can always revert later.